Refactor existing extern int const to enum classes

jmythms opened the pull request:
Changes all remaining extern int const-s to enum classes.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

jmythms committed:
broken - TARCOGLayerType

jmythms committed:
enum TARCOGLayerType

jmythms committed:
removed minlayer maxlayer

jmythms committed:
TARCOGThermalModel

jmythms committed:
rm unnecessary include

jmythms committed:
Weather typo

jmythms committed:
DeflectionCalculation

jmythms committed:
DeflectionParameters

jmythms committed:
CalcForcedVentilation

jmythms committed:
StandardRatings

jmythms committed:
fix StandardRatings Pumps

jmythms committed:
ORT stepTypeZone, HVAC

jmythms committed:
removed more extern int consts

jmythms committed:
constexpr

jmythms committed:
CLangFormat

jmythms committed:
Merge attempt - DO.hh

jmythms committed:
Merge attempt - DO.cc

jmythms committed:
Merge attempt - DXC.cc

jmythms committed:
Merge attempt - HBM.cc

jmythms committed:
Merge attempt - HVACVRF.cc

jmythms committed:
Merge attempt - OAU.hh

jmythms committed:
Merge attempt - OAU.cc

jmythms committed:
Merge attempt - Pumps.hh

jmythms committed:
Merge attempt - Pumps.cc

jmythms committed:
Merge attempt - Solver.*

jmythms committed:
Merge attempt - TARCOGArgs.*

jmythms committed:
Merge attempt - TARCOGCommon.cc

jmythms committed:
Merge attempt - TARCOGDeflection.*

jmythms committed:
Merge attempt - TARCOGMain.cc

jmythms committed:
Merge attempt - TARCOGOutput.cc

jmythms committed:
Merge attempt - TARCOGParams.*

jmythms committed:
Merge attempt - ThermalISO15099Calc.*

jmythms committed:
Merge attempt - WeatherMan.unit

jmythms committed:
Merge attempt - WindowComplexMan.hh

jmythms committed:
fix initialization Array1D

jmythms committed:
fix complex glazing

jmythms committed:
Fix map + other regression diffs

jmythms committed:
Merge branch 'develop' into externIntConsts

jmythms committed:
Merge branch 'develop' into externIntConsts

amirroth commented on src/EnergyPlus/Pumps.hh:
```diff
@@ -91,16 +91,20 @@ namespace Pumps {
+        Cond = 103,
+        Bank_VarSpeed = 104,
+        Bank_ConSpeed = 105
+    };
```
Try to make these just 1-5, instead of 101-105. See what happens. Live a little!

jmythms replied:
Will remember consistency is key!

Myoldmopar replied:
I'd say even go farther and do -1 for the invalid case and 0-4 for the valid entries.

mjwitte replied:
Just beware any if statements that may be checking for `=0` or `<0`  or `<100` etc. instead of testing against an enum name.

jmythms replied:
Thanks. Most of these comparison checks usually pop up as IDE/compiler errors when we convert the `extern int const` definitions to `enum`s, saying that enums cannot be compared with ints. I have tried to refactor any checks like `<0` and `<100` to check against the actual enumerators, but I will keep an eye out. If I did things right (fingers crossed), converting the (101-105) to (-1 to 4) should be simple and I would not have to make changes anywhere else, since the associated `int`s are not being used for comparisons anymore. Thank you for checking and the advice,

amirroth replied:
What is the i in `iPumpType`? `int`? I'm not against [Hungarian Notation](https://en.wikipedia.org/wiki/Hungarian_notation) (they made us use it at Microsoft back in the day) but I didn't think we use it in EnergyPlus. Neither here nor there.

mitchute replied:
That's probably a precedent I started or at least perpetuated. I believe I kept running into naming conflicts, so I just added the "i" to help distinguish the enum.

jasondegraw replied:
Let's please not use Hungarian Notation. There are other ways to address conflicts.

jmythms replied:
I think another small advantage being the distinction between enum class names and variables that use them (like [here](https://github.com/NREL/EnergyPlus/blob/f4a11a4781cf31579144724e559705ed1b9ad4f2/src/EnergyPlus/Pumps.hh#L148)). If there is a standard format for naming enum class names that we would like to follow, I'm glad to change to that.

amirroth replied:
So, now that I think of it this isn't really HN, because `iPumpType` is a type not a variable. If this were HN, it would be `PumpType_t`. @jasondegraw, your thoughts?

jasondegraw replied:
I'm pretty sure that a more standard approach would be to use lower case initial characters for variables (`thisIsMyVariable`) and upper case initial characters for types (`ThisIsMyType`).

Myoldmopar replied:
I support that simple style.  In this case we are already in the pump's namespace, so you could probably go pretty simple and still have clarity:

```
namespace EnergyPlus {
 namespace Pumps {
  enum class Type { Constant, Variable, Etc }; // or PumpType if you really want
  struct PumpData {
   ...
   Type type; // or myType if you want
   ...
  }
 }
}

mitchute replied:
I'll name them to whatever convention is generally agreed upon as best practice. I'm not aware of anything in our Wiki that describes this, although, those pages are somewhat out of date. I believe I was just looking at the Google C++ style guide on enumerators (https://google.github.io/styleguide/cppguide.html#Enumerator_Names) that says they can be named like constants (see here: https://google.github.io/styleguide/cppguide.html#Constant_Names)

jasondegraw replied:
@amirroth I'd take `Type_t` over `iType` any day of the week.

jasondegraw replied:
@mitchute I don't really find that to be too persuasive. I believe I've seen cases where the enumeration is getting named something like `iThingTypeNum`. It's no longer a number, so the obvious name to me would be `Thing` or `ThingType`. I've been slowly changing variable names to start with lower case as I work on objects and starting out that way with new objects, so from my point of view this method of naming is heading the wrong way.

mitchute replied:
You'll get no disagreement from me that there are plenty of bad examples of type names inside of EnergyPlus. In fact, that's probably what lead me down this road to begin with. Probably something like ```PumpType``` not being an enum type, but being a poorly named struct or local variable that was causing naming problems. This was a quick way of getting around the issue. Arguably it wasn't the best, but it was *A* way.

That being said, I don't think the leading "i" or "k" or whatever is really that bad. At least for me, since I've been working with these so much over the last year, my brain is already programmed to understand that the leading "i" means that I'm dealing with an enumerator, and that I don't have to drill in and see whether this really is an enumerator "type", or some other poorly named struct.

However, as I said, I'll do whatever.

amirroth replied:
@Myoldmopar, so with your example "PumpType" enumerations would be `Type::Constant` within the Pumps namespace and `Pumps::Type::Constant` outside of it? I don't know if I like that... 

PS I would similarly complain about the type name itself but I realize that all of that is just going to be `auto` anyway so what's the point ...

amirroth commented on src/EnergyPlus/TARCOGCommon.cc:
```diff
@@ -70,14 +70,15 @@ namespace TARCOGCommon {
+        return layertype == TARCOGLayerType::VENETBLIND_HORIZ || layertype == TARCOGLayerType::VENETBLIND_VERT ||
+               layertype == TARCOGLayerType::WOVSHADE || layertype == TARCOGLayerType::PERFORATED || layertype == TARCOGLayerType::BSDF ||
+               layertype == TARCOGLayerType::DIFFSHADE;
     }
```
Perfect spot for the `BITF` trick.

jmythms replied:
I did not understand. Do you mean bit fields?

amirroth replied:
Yes, `BITF_TEST_ANY(BITF(layertype), BITF(TARCOGLayerType::VENETBLIND_HORIZ) | BITF(TARCOGLayerType::VENETBLIND_VERT) | ... )`. One instruction instead of 12.

jmythms replied:
Awesome. Will do!

jmythms replied:
Is it recommended to use `BITF` in other places as well?

amirroth commented on src/EnergyPlus/TARCOGParams.hh:
```diff
@@ -65,39 +65,56 @@ int constexpr maxlay2(maxlay * 2);  // maximum number of glass surfaces (maxlay*
+    BSDF = 5,
+    VENETBLIND_VERT = 6
+};
 
```
So ... in some places we are setting `UNASSIGNED` as `0`, in others we are using `-1`. We should settle on one, preferably `-1` but we could make `0` work also.

jmythms replied:
Got it. Will change to `-1`, and let the group know as well 👍🏽

jmythms committed:
iPumpType int assignments

jmythms committed:
enum class HPdefrostControl

jmythms committed:
enum class DeflectionParameters

jmythms committed:
enum class CalcForcedVentilation

jmythms committed:
fix enum class CompType

jmythms committed:
checkpoint fix enum class CompType2

jmythms committed:
fix comment

jmythms committed:
BITF implementation layertype

jmythms committed:
BITF implementations more

amirroth commented on src/EnergyPlus/Pumps.cc:
```diff
@@ -2051,7 +2051,7 @@ void CalcPumps(EnergyPlusData &state, int const PumpNum, Real64 const FlowReques
 
-    } else if (PumpType == iPumpType::Bank_ConSpeed || PumpType == iPumpType::Bank_VarSpeed) {
+    } else if BITF_TEST_ANY(BITF(PumpType), BITF(iPumpType::Bank_ConSpeed) | BITF(iPumpType::Bank_VarSpeed)) {
 
```
If you have an if-else ladder of these things, you can also use switch/case:

```
switch(PumpType) {
   case iPumpType::ConSpeed:
   case iPumpType::VarSpeed:
   case iPumpType::Cond: 
        blahblahblah;
        break;
   case iPumpType::Bank_ConSpeed:
   case iPumpType::Bank_VarSpeed:
       yadayadayadah;
       break;
   default:
      thisisprobablyanerror;
      break;
}
```

jmythms replied:
Thank you. Is there a preferred/optimum way to do it (Or decide between if-else/switch/BITF methods)? In this case, I am guessing BITF doesn't give many advantages since there are just three+two items being compared. So it is best to change to switch/case with the most common cases put first?

amirroth replied:
There are actually two separate issues here. 

Issue 1. Should you use `Switch/case` or `if-else`? You should use `switch/case` if the `if-else` ladder has at least three or four cases and one of them is not dynamically dominant. There is also the limitation that `switch/case` is also for single variable equality tests, you cannot have a compound conditional in a `case`. That is the specific case here. For all other situations you either have to use `if-else` or you don't have to, but you should. This brings us to issue 2.

Issue 2. If you are using `if-else`, should you use `BITF`? The test here is how many things you are comparing against. `BITF` is three instructions including one branch for any number of comparisons. `X == A || X == B ...` is two instructions including one branch _per_ comparison. By the numbers (especially the branch numbers) you want to use `BITF` anytime you are comparing against two or more things. The performance advantage for two comparisons is admitted small, but it is there. One disclaimer is that this only true if the things you are comparing against are enumerations or other small compile-time constants.

jmythms replied:
Thank you very much again. Last question: does dynamically dominant mean that the probability of a case occurring is higher than others?

amirroth replied:
That's right. If one of the cases occurs 90% of the time, you want to use an `if-then` ladder and put that case first. The remaining cases will be a bit slower than they would have been had you used a `switch\case`, but this case will be faster and that is the one that matters for performance.

jmythms replied:
Got it. Thanks so much!

amirroth commented on src/EnergyPlus/Pumps.cc:
```diff
@@ -2041,8 +2042,7 @@ void CalcPumps(EnergyPlusData &state, int const PumpNum, Real64 const FlowReques
-    if (PumpType == iPumpType::ConSpeed || PumpType == iPumpType::VarSpeed || PumpType == iPumpType::Cond) {
-
+    if (BITF_TEST_ANY(BITF(PumpType), BITF(iPumpType::ConSpeed) | BITF(iPumpType::VarSpeed) | BITF(iPumpType::VarSpeed))) {
         VolFlowRate = state.dataPumps->PumpMassFlowRate / LoopDensity;
```
The last one should be iPumpType::Cond, no?

jmythms replied:
Yes, thanks for noticing this.

amirroth commented on src/EnergyPlus/TARCOGArgs.cc:
```diff
@@ -425,10 +425,15 @@ int ArgCheck(EnergyPlusData &state,
+                                                   BITF(TARCOGLayerType::VENETBLIND_VERT))
+
+        {
             ArgCheck = 22;
```
So, `BITF_NOT_TEST_ANY` is a bad name, a better one would be `BITF_TEST_NONE` (i.e., it matches none of the values). Alternatively you can also just use `!BITF_TEST_ANY`.

jmythms replied:
Makes sense! Thank you.

jmythms reviewed (commented):
Accidentally started reviewing my own PR.

jmythms committed:
BITF_TEST_NONE

jmythms committed:
fix iPumpType::Cond error

jmythms committed:
switch case EquipTypeNum-CompType

jmythms committed:
switch/case (CoilTypeNum)

jmythms committed:
switch/case PumpType

jmythms committed:
switch/case PumpType SetupOutputVariable

jmythms committed:
switch/case PumpType ReportPumps

jmythms committed:
Merge branch 'develop' into externIntConsts

jmythms committed:
working Clang-format

jmythms committed:
Added default error  to switch CalcOAUnitCoilComps

jmythms committed:
change enum class name pumpType

jmythms committed:
clang format pumps

jmythms committed:
enum class PumpType rename, convert to enum class PowerSizingMethod

jmythms committed:
Merge branch 'develop' into externIntConsts

Myoldmopar commented:
Looks like most of the review comments have been addressed and CI is all clean.  I haven't looked over the final changes but will be looking it over soon.  If anyone else has anything further go ahead and take a look today if you can.

Myoldmopar commented on src/EnergyPlus/DXCoils.hh:
```diff
@@ -223,14 +224,14 @@ namespace DXCoils {
-        Real64 OATempCompressorOn;         // The outdoor tempearture when the compressor is automatically turned back on,
+        Real64 HeatingCoilRuntimeFraction;                          // Run time fraction of the DX heating unit
+        int DefrostStrategy;                                        // defrost strategy; 1=reverse-cycle, 2=resistive
+        Optional<StandardRatings::HPdefrostControl> DefrostControl; // defrost control; 1=timed, 2=on-demand
```
Hmmmm, I don't really like this optional being added.  Can you explain why that was added here?

jmythms replied:
This was a mistake. It crept in while trying to refactor optional_int to optional<enum class>. I removed it and scanned through the rest of the PR to find any other instances of this as well.

Myoldmopar commented on src/EnergyPlus/HeatBalanceManager.cc:
```diff
@@ -8881,8 +8882,8 @@ namespace HeatBalanceManager {
 
-            if (state.dataHeatBal->ComplexShade(Loop).LayerType == csVenetianHorizontal ||
-                state.dataHeatBal->ComplexShade(Loop).LayerType == csVenetianVertical) {
+            if (BITF_TEST_ANY(BITF(state.dataHeatBal->ComplexShade(Loop).LayerType),
```
I still have questions about using these macros vs using constexpr functions, but they can wait for another time.  And I need to go back and redigest the discussion from when they were discussed previously before I make any further commentary.

Myoldmopar commented on src/EnergyPlus/TARCOGCommon.hh:
```diff
@@ -54,6 +54,11 @@
 #include <EnergyPlus/EnergyPlus.hh>
+#include <EnergyPlus/TARCOGParams.hh>
+
+#define BITF(B) (1 << (int(B)))
```
Is this a second definition of BITF and BITF_TEST_ANY?  If so they need to be pulled into a single location and reused.

jmythms replied:
Done 👍🏽

Myoldmopar reviewed (changes requested):
@jmythms this is a large effort, thanks!   Well over 99% of it looks great, I just had a request to get rid of, or at least explain, the addition of a new optional argument, and then determining if we duplicated the definition of BITF_TEST_ANY and BITF.

jmythms commented:
> @jmythms this is a large effort, thanks! Well over 99% of it looks great, I just had a request to get rid of, or at least explain, the addition of a new optional argument, and then determining if we duplicated the definition of BITF_TEST_ANY and BITF.

Thank you very much for reviewing. Working on this now.

jmythms committed:
Removed Optional keyword - DXC.hh

jmythms committed:
Added const to Optional where necessary - DXC.hh

jmythms committed:
removed duplicate BITF definition

jmythms committed:
clang format

jmythms committed:
Merge branch 'develop' into externIntConsts

jmythms committed:
clang format 2

jmythms committed:
Added BITF.hh + removed unnecessary headers in HBM.cc

jmythms committed:
TARCOGCommon - Replaced BITF.hh header

jmythms committed:
Ran clang format

jmythms committed:
TARCOGCommon - Fixed BITF.hh header

jmythms commented:
> I would not include `DataSurfaces.hh` anywhere you want to use `BITF`. I would move the `BITF` definitions to `General.hh` or even to a new `BITF.hh` file.

I added the file and moved the definitions [here](https://github.com/NREL/EnergyPlus/blob/44347107eb0b8a2fd24289bbfcf6c711b1e92e06/src/EnergyPlus/BITF.hh).

jmythms committed:
Fix CMakeLists.txt :alembic:

Myoldmopar commented:
This looks great to me.  CI is super happy.  This PR will help push the codebase much further in the right direction.  Thanks for pushing on this @jmythms.  I will go ahead and merge this now, and if there are any follow-up issues, let's tackle them in the next PR!

jmythms commented:
@Myoldmopar Thank you very much! 🎉
